src/chainstate/stacks/db/blocks.rs

```diff
@@ -3380,152 +3396,247 @@ impl StacksChainState {
-                    };
-
-                    if can_attach {
-                        // try and load up this staging block and its microblocks
-                        match StacksChainState::load_staging_block(
-                            blocks_conn,
-                            blocks_path,
+                        debug!(
+                            "Consider block {}/{} whose parent is {}/{}",
                             &candidate.consensus_hash,
                             &candidate.anchored_block_hash,
-                        )? {
-                            Some(staging_block) => {
-                                // must be unprocessed -- must have a block
-                                if staging_block.block_data.len() == 0 {
-                                    return Err(Error::NetError(net_error::DeserializeError(
-                                        format!(
-                                            "No block data for staging block {}",
-                                            candidate.anchored_block_hash
-                                        ),
-                                    )));
-                                }
+                            &candidate.parent_consensus_hash,
+                            &candidate.parent_anchored_block_hash
+                        );
 
-                                // find its microblock parent stream
-                                match StacksChainState::find_parent_staging_microblock_stream(
-                                    blocks_conn,
-                                    blocks_path,
-                                    &staging_block,
-                                )? {
-                                    Some(parent_staging_microblocks) => {
-                                        return Ok(Some((
-                                            parent_staging_microblocks,
-                                            staging_block,
+                        let can_attach = {
+                            if candidate.parent_anchored_block_hash == FIRST_STACKS_BLOCK_HASH {
+                                // this block's parent is the boot code -- it's the first-ever block,
+                                // so it can be processed immediately
+                                true
+                            } else {
+                                // not the first-ever block.  Does this connect to a previously-accepted
+                                // block in the headers database?
+                                let hdr_sql = "SELECT * FROM block_headers WHERE block_hash = ?1 AND consensus_hash = ?2".to_string();
+                                let hdr_args: &[&dyn ToSql] = &[
+                                    &candidate.parent_anchored_block_hash,
+                                    &candidate.parent_consensus_hash,
+                                ];
+                                let hdr_rows = query_rows::<StacksHeaderInfo, _>(
```

Comment:
This query should use `query_row` or `query_row_panic` (if you want it to panic when more than 1 row is returned).

Reply:
Done